File: framework/languages/java/e2e/ResilienceLifecycle/Client/src/main/java/systems/zlink/e2e/resiliencelifecycle/client/Scenarios/RlD2ObserverFaultScenario.java
package systems.zlink.e2e.resiliencelifecycle.client.Scenarios;

import systems.zlink.e2e.resiliencelifecycle.client.Support.ResilienceScenarioContext;
import java.time.Duration;
import systems.zlink.e2e.resiliencelifecycle.shared.Contracts;

public final class RlD2ObserverFaultScenario {
    private RlD2ObserverFaultScenario() {
    }

    public static void run(ResilienceScenarioContext context) {
        context.post(context.adminA() + "/admin/fault/observer-throws");
        context.post(context.adminB() + "/admin/fault/observer-throws");
        try {
            context.requestUnhandled("d2-observer-fault", Duration.ofSeconds(3));
            throw new IllegalStateException("RL-D2 missing handler request unexpectedly completed");
        } catch (RuntimeException expected) {
            // Observer failures are reported through structured logging. They
            // do not create an application callback or evidence DTO.
        } finally {
            context.post(context.adminA() + "/admin/fault/none");
            context.post(context.adminB() + "/admin/fault/none");
        }
        Contracts.WorkRes followUp = context.request("rl-d2-after", Duration.ofSeconds(3));
        ResilienceScenarioContext.ensure("work:rl-d2-after".equals(followUp.value()),
            "RL-D2 messaging did not continue after observer failure");
        context.waitForEvidenceAny("WorkReq", context.adminA(), context.adminB());
        System.out.println("scenario RL-D2 passed");
    }
}

File: framework/languages/java/e2e-kotlin/ResilienceLifecycle/Client/src/main/kotlin/systems/zlink/e2e/kotlin/resiliencelifecycle/Scenarios/RlD2ObserverFaultScenario.kt
package systems.zlink.e2e.kotlin.resiliencelifecycle

fun ClientScenarioContext.runObserverFaultScenario() {
    post("${adminA()}/admin/fault/observer-throws")
    post("${adminB()}/admin/fault/observer-throws")
    waitForEvidence(adminA(), "ObserverFaultMode")
    waitForEvidence(adminB(), "ObserverFaultMode")

    try {
        requestUnhandled("d2-observer-fault")
        throw IllegalStateException("RL-D2 missing handler request unexpectedly completed")
    } catch (_: RuntimeException) {
    } finally {
        post("${adminA()}/admin/fault/none")
        post("${adminB()}/admin/fault/none")
    }
    // Observer failures are written through structured logging and do not
    // create an application callback or evidence DTO.

    val followUp = requestWork("rl-d2-after")
    ensure(followUp.value() == "work:rl-d2-after", "RL-D2 follow-up payload mismatch")
    waitForEvidenceValueAny("WorkReq", "rl-d2-after", adminA(), adminB())
    println("scenario RL-D2 passed")
}
